CASSANDRA-13606 Improve handling of 2i initialization failures

bereng opened the pull request.

bereng committed:
CASSANDRA-13606 Improve handling of 2i initialization failures

bereng commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -1126,17 +1165,26 @@ public UpdateTransaction newUpdateTransaction(PartitionUpdate update, WriteConte
-        return indexers.length == 0 ? UpdateTransaction.NO_OP : new WriteTimeTransaction(indexers);
+        
+        ArrayList<Index.Indexer> idxrs = new ArrayList<>();
+        for (Index i : indexes.values())
```
Removing streams from the hot path.

bereng commented on src/java/org/apache/cassandra/index/internal/CassandraIndex.java:
```diff
@@ -689,33 +705,42 @@ private boolean isPrimaryKeyIndex()
-
-        try (ColumnFamilyStore.RefViewFragment viewFragment = baseCfs.selectAndReference(View.selectFunction(SSTableSet.CANONICAL));
-             Refs<SSTableReader> sstables = viewFragment.refs)
+        try
```
All these changes are just a try-catch sandwiching to change `supportedLoads` to NONE on exceptions.

bereng commented:
- j11 [CI](https://app.circleci.com/pipelines/github/bereng/cassandra/11/workflows/34a10f44-be84-423b-82f7-89cbb756e8ca)
- j8 [CI](https://app.circleci.com/pipelines/github/bereng/cassandra/11/workflows/2134990d-09aa-43c2-a2bc-ef3a26d4c8e2)

adelapena commented on src/java/org/apache/cassandra/index/internal/CassandraIndex.java:
```diff
@@ -146,6 +146,22 @@ protected abstract ByteBuffer getIndexedValue(ByteBuffer partitionKey,
+        switch (load)
+        {
+            case ALL:
+                return supportedLoads.equals(Loads.ALL);
```
Being an enum, we can use `==` instead of `equals`

adelapena commented on test/unit/org/apache/cassandra/index/internal/CustomCassandraIndex.java:
```diff
@@ -94,6 +95,21 @@ protected boolean supportsOperator(ColumnMetadata indexedColumn, Operator operat
+        switch (load)
+        {
+            case ALL:
+                return supportedLoads.equals(Loads.ALL);
```
Same as before, we can use `==`

adelapena commented on src/java/org/apache/cassandra/index/Index.java:
```diff
@@ -136,6 +135,7 @@
  */
 public interface Index
 {
+    public static enum Loads {READS, WRITES, ALL, NONE};
```
Using a plural name for a enum feels a little bit odd to me, what about something like `LoadType` or the more operation `OperationType`? Also I can't think of other future values for the enum given it's current usage, so I'm wondering if we could replace it by two booleans, implicit through a pair of `supportsWrites`/`supportsReads` methods. WDYT?

bereng replied:
I think it's a personal style thing. Rather than having 2 booleans and having to reason about them, the Enum makes it more 'in your face' imo: `ALL` is clear what it means vs 2 booleans etc. :shrug:

Also maybe I have too much imagination, but I can think of 'DELETE' as a load. You could have 'append-only' indexes (think banking i.e.) in the future that don't support deletes. Maybe premature optimization is being my enemy here... Or sthg with statics for custom indexes...

I prefer you decide what is best here :-). I'll rename to singulars meanwhile in case we kept that option.

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -1126,17 +1165,26 @@ public UpdateTransaction newUpdateTransaction(PartitionUpdate update, WriteConte
-
-        return indexers.length == 0 ? UpdateTransaction.NO_OP : new WriteTimeTransaction(indexers);
+        
+        ArrayList<Index.Indexer> idxrs = new ArrayList<>();
```
Ignorable nit: perhaps we could name this variable `indexers`, and use `idxrs` for the array below. Also, I wonder if it would make sense to use an iterable instead of an array in `WriteTimeTransaction`, so we can directly pass this `ArrayList` without the `toArray` conversion. Would it hurt performance?

bereng replied:
I was trying to get all this path without iterators which would be ideal, but `indexes.values()` is a `Collection` which I can't access via `get()`. The previous streams would have the same array conversion, which is suspicious as the immediate solution would be to use the `ArrayList`. Being the hot path maybe we want to keep the 'pure' array index access even though `ArrayLists`'s overhead might be minimal if anyting at all. Sounds good?

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -424,16 +427,47 @@ public static String getIndexName(String cfName)
      * @param sstables      the SSTables to be (re)indexed
      * @param indexes       the indexes to be (re)built for the specifed SSTables
      * @param isFullRebuild True if this method is invoked as a full index rebuild, false otherwise
+     * @param isFullRebuild True if this method is invoked as an index rebuild or recovery
```
```suggestion
     * @param isRecovery {@code true} if this method is invoked as an index rebuild or recovery
```

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -424,16 +427,47 @@ public static String getIndexName(String cfName)
     }
+    
+    /**
+     * Does a blocking full rebuild/reovery of the specifed indexes from all the sstables in the base table.
```
```suggestion
     * Does a blocking full rebuild/recovery of the specifed indexes from all the sstables in the base table.
```

bereng commented:
- [Ci j11](https://app.circleci.com/pipelines/github/bereng/cassandra/15/workflows/2c0458de-6534-47e6-8e7c-4909e7827001)
- [CI j8](https://app.circleci.com/pipelines/github/bereng/cassandra/15/workflows/fdab9b26-b68c-41a8-8fbe-c739fff9f37f)

^are these runs good enough or do you recommend I run any other dtests? Thx in advance

bereng commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -485,14 +471,15 @@ private void buildIndexesBlocking(Collection<SSTableReader> sstables, Set<Index>
             Map<Index.IndexBuildingSupport, Set<Index>> byType = new HashMap<>();
             for (Index index : indexes)
             {
+                boolean isRecovery = !index.supportsLoad(Loads.ALL);
```
This is the only contentious point imo. I don't know if we should call a `Index#isRecoveryNeeded()` method instead rather than trying to infer it at the SIM level. The former assumes `ALL` is mandatory for an index to the 'right', the latter allows for the index implementation to decide that. Both make sense... It's our choice tbh and switching from one to the other in the future it's easy enough anyway.

adelapena commented on src/java/org/apache/cassandra/index/Index.java:
```diff
@@ -136,6 +135,7 @@
  */
 public interface Index
 {
+    public static enum LoadType {READ, WRITE, ALL, NONE};
```
+1 to the new enum names. It would be nice to add a very brief comment about what it is, for the sake of the unaware reader. The static and the semicolon are not needed:
```suggestion
    public enum LoadType {READ, WRITE, ALL, NONE}
```

bereng replied:
Argh! you're SO right! My bad apologies.

adelapena commented on src/java/org/apache/cassandra/index/internal/CassandraIndex.java:
```diff
@@ -146,6 +146,22 @@ protected abstract ByteBuffer getIndexedValue(ByteBuffer partitionKey,
+    
+    public boolean supportsLoad(LoadType load)
+    {
+        switch (load)
+        {
+            case ALL:
+                return supportedLoads == LoadType.ALL;
+            case READ:
+                return supportedLoads == LoadType.ALL || supportedLoads == LoadType.READ;
+            case WRITE:
+                return supportedLoads == LoadType.ALL || supportedLoads == LoadType.WRITE;
+            default:
+                return false;
+        }
```
This block of code is repeated in a couple of test index implementations, and it's to expect that other implementations would need to repeat it. I think that defining that `ALL` includes `READ` and `WRITE`, etc. seems more a characteristic of `LoadType` than the index itself, so perhaps we could move this logic to the enum itself, for example:
```java
public enum LoadType {
    READ, WRITE, ALL, NONE;

    public boolean accepts(LoadType load)
    {
        switch (this)
        {
            case ALL:
                return true;
            case READ:
                return load == LoadType.READ || load == LoadType.NONE;
            case WRITE:
                return load == LoadType.WRITE || load == LoadType.NONE;
            default:
                return false;
        }
    }
}
```
That way we would simplify usage later:
```java
@Override
public boolean supportsLoad(LoadType load)
{
    return supportedLoadType.accepts(load);
}
```

bereng replied:
Are you saying here `READ` accepts `NONE` ?

bereng replied:
Fixed to my discretion. Feel free to reopen if I got it wrong.

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -338,20 +355,20 @@ public void markAllIndexesRemoved()
+               }
+
+               buildIndexesBlocking(allSSTables, toRebuild, true);
+           }
```
Nit: this seems to have missed alignment with an extra tab

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -424,10 +441,13 @@ public static String getIndexName(String cfName)
+     * Performs a blocking (re)indexing/recovery of the specified SSTables for the specified indexes.
      *
+     * If the index doesn't support ALL {@link Index.LoadType} it performs a recovery {@link getRecoveryTaskSupport()}
+     * instead of a build {@link getBuildTaskSupport()}
```
```suggestion
     * If the index doesn't support ALL {@link Index.LoadType} it performs a recovery {@link Index#getRecoveryTaskSupport()}
     * instead of a build {@link Index#getBuildTaskSupport()}
```

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -619,8 +642,16 @@ private synchronized void markIndexesBuilding(Set<Index> indexes, boolean isFull
+        if (isFullRebuild && index.supportsLoad(Index.LoadType.WRITE))
+        {
+            writableIndexes.add(indexName);
+            logger.info("Index [" + indexName + "] became writable.");
```
I think that `became` suggests a change of status, as if it weren't writable before. We could either say that is writable without telling if it was so before or, even better, use the boolean returned by `queryableIndexes.add`:
```suggestion
            if (writableIndexes.add(indexName))
                logger.info("Index [" + indexName + "] became writable.");
```
The same applies to queryable indexes above.

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -1126,17 +1158,21 @@ public UpdateTransaction newUpdateTransaction(PartitionUpdate update, WriteConte
+        ArrayList<Index.Indexer> idxrs = new ArrayList<>();
+        for (Index i : indexes.values())
+        {
+            Index.Indexer idxr = i.indexerFor(update.partitionKey(), update.columns(), nowInSec, ctx, IndexTransaction.Type.UPDATE);
+            if (idxr != null && isIndexWritable(i))
```
We could check if the index is writable before creating the indexer:
```suggestion
            if (!isIndexWritable(i))
                continue;

            Index.Indexer idxr = i.indexerFor(update.partitionKey(), update.columns(), nowInSec, ctx, IndexTransaction.Type.UPDATE);
            if (idxr != null)
```
The same applies to the other usages of `isIndexWritable` in `IndexGCTransaction#commit()` and `CleanupGCTransaction#commit()`. Alternatively, we could transform the set `SIM#writableIndexes` into a `Map<String, Index>`, and iterate its values this instead of `SIM#indexes` values in the three places. That way we wouldn't even need the `isIndexWritable` method.

bereng replied:
Sgtm. That was my first implementation I can't remember for the sake of me why I changed it to a set.

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -1356,7 +1396,7 @@ public void commit()
                 {
                     Index.Indexer indexer = index.indexerFor(key, columns, nowInSec, ctx, Type.COMPACTION);
-                    if (indexer == null)
+                    if (indexer == null || !indexManager.isIndexWritable(index))
```
See comment above about either calling `isIndexWritable` before creating the indexer, or directly passing `writableIndexes` to the `IndexGCTransaction ` constructor, so we don't need to this check.

adelapena commented on src/java/org/apache/cassandra/index/internal/CassandraIndex.java:
```diff
@@ -79,6 +78,7 @@
     protected ColumnFamilyStore indexCfs;
     protected ColumnMetadata indexedColumn;
     protected CassandraIndexFunctions functions;
+    protected LoadType supportedLoads = LoadType.ALL;
```
Nit:
```suggestion
    protected LoadType supportedLoadType = LoadType.ALL;
```

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1052,6 +1052,58 @@ public void testIndexQueriesWithIndexNotReady() throws Throwable
+
+        ReadOnlyIndex index = (ReadOnlyIndex) getCurrentColumnFamilyStore().indexManager.getIndexByName(indexName);
+        assertEquals(0, index.rowsInserted.size());
+        execute("DROP index " + KEYSPACE + "." + indexName);
```
We could try a recovery here, and verify that the index starts to accepts writes after that recovery. Also, in `ReadOnlyIndex`, we could throw an exception in `getInitializationTask`, and provide a mock implementation `getRecoveryTaskSupport` that enables writes. Note that forcing an exception during initialization would make us to miss the call to `markIndexBuilt`, so the index won't be added to `SIM#queryableIndexes` and it won't accepts reads independently of its `ReadOnlyIndex#supportsLoad` implementation. That's why I think we need some new logic in `SIM#markIndexFailed` too.

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1052,6 +1052,58 @@ public void testIndexQueriesWithIndexNotReady() throws Throwable
+        }
+        finally
+        {
+            execute("DROP index " + KEYSPACE + "." + indexName);
```
Similarly to the comment before, we could try a recovery, and simulate the initializitation failure throwing an exception in `WriteOnlyIndex#getInitializationTask`.

adelapena commented on test/unit/org/apache/cassandra/index/SecondaryIndexManagerTest.java:
```diff
@@ -119,37 +124,47 @@ public void addingSSTablesMarksTheIndexAsBuilt() throws Throwable
 
     @Test
-    public void cannotRebuildWhileInitializationIsInProgress() throws Throwable
+    public void cannotRebuilRecoverdWhileInitializationIsInProgress() throws Throwable
```
```suggestion
    public void cannotRebuildRecoverWhileInitializationIsInProgress() throws Throwable
```

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1375,9 +1427,10 @@ public void testIndexOnFrozenUDT() throws Throwable
 
         execute("INSERT INTO %s (k, v) VALUES (?, ?)", 0, udt1);
         String indexName = createIndex("CREATE INDEX ON %s (v)");
+        assertTrue(waitForIndex(keyspace(), tableName, indexName));
```
I think that this is an important change of behaviour in indexes, they are missing writes done during their initialization. I think that this problem can be solved if we add them to `SIM#writableIndexes` at the beginning of their registration (before initialization), and we remove them from `writableIndexes` in `SIM#markIndexFailed`. WDYT?

bereng replied:
I think this is a feature and the whole point of the PR. Not getting writes until the index is ready. But having said that we don't have any commit log now that we changed to ignoring writes instead of failing them. So the only option left is what you say: we forward the writes to the index and he'll have to sort it out... Still it irks me and this behavior is not symmetrical  to the read path :shrug:

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1560,4 +1613,77 @@ public IndexBlockingOnInitialization(ColumnFamilyStore baseCfs, IndexMetadata in
+    }
+
+    /**
+     * <code>CassandraIndex</code> that only supports writes. Could be intentional or a result of a bad init
```
Nit: is not a `CassandraIndex`:
```suggestion
     * {@code StubIndex} that only supports writes. Could be intentional or a result of a bad init.
```

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1560,4 +1613,77 @@ public IndexBlockingOnInitialization(ColumnFamilyStore baseCfs, IndexMetadata in
+    }
+
+    /**
+     * <code>CassandraIndex</code> that only supports reads. Could be intentional or a result of a bad init
```
Nit: it's not a `CassandraIndex `:
```suggestion
     * {@code StubIndex} that only supports reads. Could be intentional or a result of a bad init.
```

adelapena commented on test/unit/org/apache/cassandra/index/SecondaryIndexManagerTest.java:
```diff
@@ -719,4 +746,20 @@ public boolean shouldBuildBlocking()
+
+        public boolean supportsLoad(LoadType load)
+        {
+            return load.equals(LoadType.READ);
```
```suggestion
            return load == LoadType.READ;
```

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1560,4 +1613,77 @@ public IndexBlockingOnInitialization(ColumnFamilyStore baseCfs, IndexMetadata in
+
+        public boolean supportsLoad(LoadType load)
+        {
+            return load.equals(LoadType.READ);
```
```suggestion
            return load == LoadType.READ;
```

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1560,4 +1613,77 @@ public IndexBlockingOnInitialization(ColumnFamilyStore baseCfs, IndexMetadata in
+
+        public boolean supportsLoad(LoadType load)
+        {
+            return load.equals(LoadType.WRITE);
```
```suggestion
            return load == LoadType.WRITE;
```

adelapena commented on test/unit/org/apache/cassandra/index/internal/CustomCassandraIndex.java:
```diff
@@ -77,6 +77,7 @@
     protected ColumnFamilyStore indexCfs;
     protected ColumnMetadata indexedColumn;
     protected CassandraIndexFunctions functions;
+    protected LoadType supportedLoads = LoadType.ALL;
```
```suggestion
    protected LoadType supportedLoadType = LoadType.ALL;
```

adelapena commented on test/unit/org/apache/cassandra/index/SecondaryIndexManagerTest.java:
```diff
@@ -719,4 +746,20 @@ public boolean shouldBuildBlocking()
     }
+
+    /**
+     * <code>CassandraIndex</code> that only supports reads. Could be intentional or a result of a bad init
```
```suggestion
     * <code>TestingIndex</code> that only supports reads. Could be intentional or a result of a bad init
```

adelapena reviewed (changes requested):
This is looking good. My major concern is that indexes seem to be missing writes done during initialization. This happens because they are not included into `SIM#writableIndexes` until `SIM#markIndexBuilt` is called. I think we should add indexes to `writableIndexes` as soon as they are registered, and remove them in `markIndexFailed`. Conversely, recovering an index would optimistically mark it as writable as soon as the task starts, setting it as not writable only if it fails, so we don't miss writes done while the recovery task is running. 

That way, the meaning of "unavailable for writes" would be that we skip writes as an optimization when we know that we are going to need to do them again during recovery.

Also, the implementation of `Index#supportsLoad` is based on an internal state that index implementations manage with try/catch blocks in their initialization task. I think that most of that can be done in the SIM, replacing `Index#supportsLoad(LoadType)` by a simpler `Index#supportedLoadWhenFailed` getter, returning the `LoadType` that is supported when the index is failed. That way, `SIM#markIndexBuilt` would take care of re-adding the index to `writable/queryableIndexes` in recoveries, and `SIM#markIndexFailed` would be the only caller of `Index#supportedLoadWhenFailed`. If we wanted more flexibility in what operations are supported during the initial build, we could always have an `Index#supportedLoadWhenCreating`.

What do you think? Does it make any sense?

bereng commented:
- [CI j11](https://app.circleci.com/pipelines/github/bereng/cassandra/18/workflows/dcb313cc-b897-4a19-8840-09550a62d6d2)
- [CI j8](https://app.circleci.com/pipelines/github/bereng/cassandra/18/workflows/c3f34b71-5319-4a4b-b6b0-24d90462c71a)

@adelapena 2 things:
- I modified the enum logic
- I didn't quite get the `supportedLoadWhenFailed` suggestion, neither I suspect would bring much value? but I am probably missing sthg. Please feel free to push or rephrase if you feel strongly about it.

Also, reminder to myself to rebase, squash and run a full dtest suite once we're happy.

adelapena commented on src/java/org/apache/cassandra/index/Index.java:
```diff
@@ -136,6 +135,28 @@
+                case ALL:
+                    return true;
+                case READ:
+                    return load == LoadType.READ;
```
That's tricky because it's not even going to happen, I understand that if the type of load of an index is `READ` or `WRITE` it should accept `NONE`, meaning that a read/write accepts a no-op workload. Perhaps it would be easier to replace the method by a pair of `acceptsReads/acceptsWrites` methods on the enum matching the expected usage, especially if we replace `Index#supportsLoad(LoadType)` by a simpler `supportedLoadWhenFailed()` getter (more details on this later).

bereng replied:
Aha, this is where I see the diff povs were tackling this problem from. You see it as 'accepts' and I see it as 'I am ready for'. Ofc NONE makes no sense to me, but it does to you...
- I accept a no-op load
- I am ready for NONE loads
:thinking:

bereng replied:
I could rename `NONE` to `NO-OP` to align to your train of thought. Otherwise `NONE` sounds weird to me.

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -451,15 +475,30 @@ private void buildIndexesBlocking(Collection<SSTableReader> sstables, Set<Index>
+                    rebuildOrRecoveryTask = (cfs, idxs, sstbls) ->
+                    {
+                        for (Index i : idxs)
+                            if (writableIndexes.put(i.getIndexMetadata().name, i) == null)
```
At this point the sstable references would have already been acquired by `SIM#rebuildIndexesBlocking`, so I think we might lose writes in between. If this is true we would probably need to optimistically mark them as writable in `rebuildIndexesBlocking` before acquiring the sstables, as we do with the initial build. WDYT?

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -619,8 +658,21 @@ private synchronized void markIndexesBuilding(Set<Index> indexes, boolean isFull
+        if (isFullRebuild && index.supportsLoad(Index.LoadType.READ))
+        {
             queryableIndexes.add(indexName);
+            logger.info("Index [" + indexName + "] became queryable.");
```
That could be logged only if the index wasn't queryable before.

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -619,8 +658,21 @@ private synchronized void markIndexesBuilding(Set<Index> indexes, boolean isFull
+        }
+        if (isFullRebuild && index.supportsLoad(Index.LoadType.WRITE))
+        {
+            if (writableIndexes.put(indexName, index) == null)
```
If we are going to optimistically mark the indexes as writable because of the risk of missing writes done during the index build, at this point `writableIndexes` will always contain the index, I think. Perhaps we should only mark the indexes as writable (and log) only at the beginning of initial build and recovery, and unmark them (and log) only in `markIndexFailed`. Queryability doesn't need optimistic marking so it's fine at it is.

bereng replied:
I understand what you say. But this _not_ being in the hot path I think that optimization I would avoid. That would leave the method inherently wrong as it doesn't do what the method is for: mark the index as built. It rather relies on external methods doing that for him. That is not ideal imo future proof wise were others might call `markIndexBuilt` and have to take that into account.

Unless you feel strongly about it I'd favor leaving it as it is.

adelapena replied:
Makes sense to me, it's better to have the writable management also here. Independently, I would find it slightly more readable with some nesting:
```java
if (isFullRebuild)
{
    if (index.supportsLoad(Index.LoadType.READ))
    {
        if (queryableIndexes.add(indexName))
            logger.info("Index [" + indexName + "] became queryable.");
    }
    
    if (index.supportsLoad(Index.LoadType.WRITE))
    {
        if (writableIndexes.put(indexName, index) == null)
            logger.info("Index [" + indexName + "] became writable.");
    }
    else
    {
        writableIndexes.remove(indexName);
        logger.info("Index [" + indexName + "] is non-writable.");
    }
}
```
Feel free to ignore if you don't agree.

bereng replied:
lgtm +1

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -1148,7 +1204,7 @@ public CompactionTransaction newCompactionTransaction(DecoratedKey key,
     {
         // the check for whether there are any registered indexes is already done in CompactionIterator
-        return new IndexGCTransaction(key, regularAndStaticColumns, keyspace, versions, nowInSec, listIndexes());
+        return new IndexGCTransaction(key, regularAndStaticColumns, keyspace, versions, nowInSec, writableIndexes.values());
```
Nice :)

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1052,6 +1055,104 @@ public void testIndexQueriesWithIndexNotReady() throws Throwable
+            getCurrentColumnFamilyStore().indexManager.rebuildIndexesBlocking(ImmutableSet.of(indexName));
+            execute("INSERT INTO %s (pk, ck, value) VALUES (?, ?, ?)", 2, 1, 1);
+            assertEquals(1, index.rowsInserted.size());
+            execute("DROP index " + KEYSPACE + "." + indexName);
```
Nit: we can use the provided `dropIndex` method:
```suggestion
            dropIndex("DROP INDEX %s." + indexName);
```
There are three more usages below.

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1052,6 +1055,104 @@ public void testIndexQueriesWithIndexNotReady() throws Throwable
+            catch (IndexNotAvailableException e)
+            {
+                assertTrue(true);
+            }
```
Nit: we can use the provided `assertInvalidThrow`:
```suggestion
            assertInvalidThrow(IndexNotAvailableException.class, "SELECT value FROM %s WHERE value = 1");
```

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1375,9 +1476,10 @@ public void testIndexOnFrozenUDT() throws Throwable
 
         execute("INSERT INTO %s (k, v) VALUES (?, ?)", 0, udt1);
         String indexName = createIndex("CREATE INDEX ON %s (v)");
+        assertTrue(waitForIndex(keyspace(), tableName, indexName));
```
We don't need these changes in the usage of `waitForIndex` anymore since we haven't changed the standard index behaviour.

adelapena reviewed (commented):
> I didn't quite get the `supportedLoadWhenFailed` suggestion, neither I suspect would bring much value? but I am probably missing sthg. Please feel free to push or rephrase if you feel strongly about it.

I'll try to explain myself better. Currently we are managing the writability/queryability status of each index in two separate places:
* In the SIM, with the `writableIndexes/queryableIndexes` collections.
* In the `Index` instances, that hold an instance of `LoadType`, manipulated in the build tasks and exposed through the `Index#supportsLoad` method.

I think that this duplicate status management can be simplified, so the SIM would be the only responsible for the writability/queryability status management, and the indexes would merely express their capabilities/preferences when the build has failed. To do so:
* For `Index`, instead of holding a `LoadType` instance updated with a try-catch during its build, we would just have an `Index#supportedLoadTypeWhenNotBuilt()` (or similar name) getter method that would return the `LoadType` that the index supports in case of not being properly build. That method **would always return the same value**, independently of the sate of the index, because it would only describe a capability of the index implementation.
* `SecondaryIndexManager` would have all the information regarding index status, so it wouldn't need to query the indexes by their internal status. Instead, it would just ask the indexes: "what kind of loads do you support in case you are not completely built?". This way, the SIM status management would more or less be:
  * On starting index initial build or recovery, always add the index to `writableIndexes` and perhaps remove it from `queryableIndexes` if the index `LoadType`-when-unbuilt policy says so.
  * On `markIndexBuilt`, always add the index to both `writableIndexes` and `queryableIndexes`.
  * On `markIndexFailed`, remove the index from `writableIndexes` if its `LoadType`-when-unbuilt policy says so.

The main benefit of this would be simplifying writability/queryability status management, while minimizing the index implementation responsibilities. 

We of course would lose some flexibility in terms of implementations that decide to be always read-only or write-only, or that can accept different loads depending on the type of build failure, etc. But I don't think we currently have a use case for them, and I'd prefer to focus in what we strictly need. In any case, I think it wouldn't be hard for those hypothetical implementations to do all these things without specific SIM support.

What do you think?

bereng commented:
@adelapena thanks for taking the time to rephrase that bit about the indexes and the load type.

You're mainly replacing the supported load type variable in `Index` with a method call and removing that state. Notice `CassandraIndex` defaults to `ALL` for the state and the `Index#supportsLoad()` method defaults to always `true` except for `NONE`. I would argue that makes any new index implementation concerns for that a zero overload. Like you can ignore it as defaults are already reflecting the std behavior.

So given the above it boils down to me seeing this as a feature and you probably as an unnecessary optimization. I'd obviously like to leave it as it is (but I am the father of the baby lol) but you know the codebase an it's uses better. So let me know if you'd like me to change it please?

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1052,6 +1055,88 @@ public void testIndexQueriesWithIndexNotReady() throws Throwable
+
+            // On bad init writes are not forwarded to the index
+            index.reset();
+            LoadTypeConstrainedIndex.setSupportedLoad(LoadType.WRITE);
```
Shouldn't this use `LoadType.READ`, since we are testing the read only index?

bereng replied:
Right, but you don't know if you're not forwarding inserts bc of a bad init or bc writes are not supported. But as we test that case in the next write-centic method we can do reads here.

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1052,6 +1055,88 @@ public void testIndexQueriesWithIndexNotReady() throws Throwable
+            index.reset();
+            getCurrentColumnFamilyStore().indexManager.rebuildIndexesBlocking(ImmutableSet.of(indexName));
+            execute("INSERT INTO %s (pk, ck, value) VALUES (?, ?, ?)", 2, 1, 1);
+            assertEquals(1, index.rowsInserted.size());
```
Could we verify reads after rebuild too, with `execute("SELECT value FROM %s WHERE value = 1")`, just in case?

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1052,6 +1055,88 @@ public void testIndexQueriesWithIndexNotReady() throws Throwable
+            // On bad init writes are not forwarded to the index
+            index.reset();
+            LoadTypeConstrainedIndex.setSupportedLoad(LoadType.WRITE);
+            index.failInit = true;
```
I think that here we are failing the initialization of the index instance that we have just removed. The index that is created right after is a new instance with `failInit = true`, so the initialization doesn't fail, and the number of insertions belongs to the deleted index.  The same applies to `testWriteOnlyIndex` below.

bereng replied:
ops! apologies.

adelapena commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -1052,6 +1055,88 @@ public void testIndexQueriesWithIndexNotReady() throws Throwable
+            indexName = createIndex("CREATE CUSTOM INDEX ON %s (value) USING '" + LoadTypeConstrainedIndex.class.getName() + "'");
+            assertTrue(waitForIndexBuilds(keyspace(), indexName));
+            execute("INSERT INTO %s (pk, ck, value) VALUES (?, ?, ?)", 1, 1, 1);
+            assertEquals(0, index.rowsInserted.size());
```
We probably should also check read queries in this case too, and also in the equivalent case `testWriteOnlyIndex`.

adelapena commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -451,15 +480,17 @@ private void buildIndexesBlocking(Collection<SSTableReader> sstables, Set<Index>
             for (Index index : indexes)
             {
-                Set<Index> stored = byType.computeIfAbsent(index.getBuildTaskSupport(), i -> new HashSet<>());
+                IndexBuildingSupport rebuildOrRecoveryTask = index.supportsLoad(LoadType.ALL) ? index.getBuildTaskSupport()
```
If I'm understanding it right, the method `Index#supportsLoad(LoadType)` (in contrast to a simpler stateless `Index#getSupportedLoadOnFailure()`) gives indexes the ability to be read-only or write-only even when their builds are successful. In that case I think that we can't use `supportsLoad` to determine whether they need a recovery task or a simple rebuild, because they will always use the recovery task. Perhaps we can use `needsFullRebuild` instead?

bereng replied:
I my view, an index that doesn't support `ALL` should be `recovered` always. But given you've put in the time to implement your approach we can abandon this.

adelapena reviewed (commented):
> You're mainly replacing the supported load type variable in Index with a method call and removing that state. Notice CassandraIndex defaults to `ALL` for the state and the `Index#supportsLoad()` method defaults to always true except for `NONE`. I would argue that makes any new index implementation concerns for that a zero overload. Like you can ignore it as defaults are already reflecting the std behavior.

Indeed new index implementations don't need to do anything if the default behaviour is always accepting both reads and writes, independently of whether the index is properly built or not. I'm not sure that's the ideal default behaviour, especially when we are shipping a different behaviour for the standard indexes. Now, if we wanted to change the default behaviour, we would need to manage the their internal state in the index implementations, as we do in `CassandraIndex`.

In contrast, using a simple getter to retrieve the `LoadType` would make it easy for implementations to define their behaviour, independently of which default behaviour we have chosen. It will of course come at the cost of losing the feature of allowing indexes to be completely read-only of write-only. As I said, I'm not sure how useful is that feature, and it seems to me that nevertheless it would quite trivial for indexes to implement such behaviour without any specific API support. 

I have put [here](https://github.com/apache/cassandra/compare/trunk...adelapena:d055a97807ca973d15bd36bf90ab2706da4edd95) a draft of what the alternative API would look like, in case you want to take a look. 

Said that, and although it would not appear to be the case, I'm not super strong on this, and I'll be willing to keep the current method I you think that there might be cases where it can be useful.

I'm running dtests [here](https://ci-cassandra.apache.org/view/patches/job/Cassandra-devbranch-dtest/111/). Particularly, [secondary_indexes_test](https://github.com/apache/cassandra-dtest/blob/master/secondary_indexes_test.py) contains tests about building status and queryability after build failure, node restart, etc. We should check that those still work, and see if we want to extend them in some way.

bereng commented:
@adelapena again, thanks a lot for looking into this. Given you've put the time into implementing your approach I have patched and adopted it. At the end of the day you're more familiar with the code and it's uses than me :-)

The dtests thing... mmmm we should try to have as much utest and as little dtest as possible imo. Let me go through them, as they're new to me, and get back to you.

Btw your dtest run looks good. I guess we didn't break much stuff :-)

Edit: On the recent 4.0 status email do you have any strong opinions on this ticket? I don't expect dtests will need much and we're close to merging it seems. I would vote to leave this in 4.0 being so close. My 2cts.

adelapena commented:
> @adelapena again, thanks a lot for looking into this. Given you've put the time into implementing your approach I have patched and adopted it. At the end of the day you're more familiar with the code and it's uses than me :-)

Great, have you pushed the merge?

> The dtests thing... mmmm we should try to have as much utest and as little dtest as possible imo. Let me go through them, as they're new to me, and get back to you

Right, I was thinking of testing the log messages informing about when an index becomes queryable/writable. The dtest [test_failing_manual_rebuild_index](https://github.com/apache/cassandra-dtest/blob/master/secondary_indexes_test.py#L327) looks ideal for this purpose, we could easily add `node.grep_log` checks after each index (re)build to check that it behaves at is supposed. I don't know if we have tools to verify logs and simulate node restarts in utest.

> Btw your dtest run looks good. I guess we didn't break much stuff :-)

This is good news, it seems that at least we didn't make it worse than it was :smiley:

> Edit: On the recent 4.0 status email do you have any strong opinions on this ticket? I don't expect dtests will need much and we're close to merging it seems. I would vote to leave this in 4.0 being so close. My 2cts.

I think we are close to merge. Extending dtests should be a quite small task, probably just the few log checks mentioned above. It's true that the ticket doesn't solve any serious problem that we have with current implementations, but it's not very invasive and it's almost done, so I would also be in favour of keeping it in 4.0

bereng commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -655,6 +695,9 @@ private synchronized void markIndexFailed(Index index)
             needsFullRebuild.add(indexName);
+
+            if (!index.getSupportedLoadTypeOnFailure().supportsWrites() && writableIndexes.remove(indexName) != null)
+                logger.info("Index [" + indexName + "] became not-writable.");
```
@adelapena please notice I changed sthg from your patch here. Specifically the queryable removal upon failure as it would fail [this](https://github.com/apache/cassandra-dtest/pull/69/files#diff-fb360a29e46831ca27d4bb8e77059503L359) which looks legit.

adelapena replied:
Perhaps the problem was assigning `LoadType.NOOP` as the default behaviour. Indeed, removing that line leaves `LoadType#supportsReads()` without usages. I think that, to preserve the behaviour of the default indexes, and to have more flexibility, we should keep that line and make a distinction between failures in the initial build task and failures in rebuilds. As I mention on the dtest review, the classic behaviour would be:
```java
/**
 * Returns the type of operations supported by the index in case its building has failed and it's needing recovery.
 * 
 * @param isInitialBuild {@code true} if the failure is for the initial build task on index creation, {@code false}
 * if the failure is for a rebuild or recovery.
 */
default LoadType getSupportedLoadTypeOnFailure(boolean isInitialBuild)
{
    return isInitialBuild ? LoadType.WRITE : LoadType.ALL;
}
```
I think that each behaviour has advantages and disadvantages depending on the use case and, while I'm quite happy with the changes in the index API, I'm a bit afraid of changing the behaviour of the standard index implementation. WDYT?

bereng replied:
Hi @adelapena thanks again for looking into this. You're raising the point I wanted to discuss with you once this ticket was merged: Is it reasonable a 'not ready' index should be serving reads or writes?

I think we're finding ourselves battling that, but that's a bigger discussion. Maybe in some other ticket we could recover my approach which iirc would defer to the index implementation what to do :shrug:

adelapena replied:
> Is it reasonable a 'not ready' index should be serving reads or writes?

For initial build it should't serve reads, but for rebuilds it's more complicated because the index could still be in good condition for most reads. IMO if we were creating the index API from scratch or adding a new implementation probably the best option would be to set not properly (re)built indexes as unable to serve reads and writes. But, given that we have had the cf-based implementation around for a while, I think we should by now preserve their original behaviour and provide the API mechanisms to change that in new implementations, and perhaps also in the standard indexes in the future. That and the recovery task seem a pair of nice improvements to ship with this ticket.

Accepting reads and writes when the rebuild has failed makes some sense in the particular case of cf-based regular indexes because they don't have an initialization task and they are based on idempotent operations on the underlaying column family. I'd say they focus in availability over consistency because even failed rebuilds always leave the index is same or better condition than before. In contrast, it's easy to imagine other implementations where a failed rebuild leaves the index completely corrupt and unusable. I think there could easily be use cases out there relying on this behaviour, and I'm afraid of arbitrarily changing that behaviour without coming back. If we still want to move to the new approach, either in this ticket or in a dedicated one, we should probably open a discussion on the mail list to see if there are deployments relaying on the classic behaviour. 

Another tempting option to ease the migration of cf-based indexes to the new consistency-over-availability approach would be making the enum returned by `getSupportedLoadTypeOnFailure` configurable through index options, for example:
```
CREATE INDEX my_idx ON t WITH OPTIONS = {'supported_load_on_failed_init': 'WRITE', 'supported_load_on_failed_rebuild': 'ALL'}
```
This is something that perhaps we could do in the followup ticket to change the default behaviour of regular cf-based indexes, while keeping this ticket focused on the API changes and the recovery task.

WDYT? Maybe I'm wrong and there isn't anyone out there relying on the old 2i behaviour and we should go straight to the the new failed-rebuild-means-broken behaviour.

bereng replied:
Hi @adelapena, regarding `return isInitialBuild ? LoadType.WRITE : LoadType.ALL;`

`isInitialBuild` is a new variable we'd have to create and track in SIM. And then `Index` would have to render which load is supported depending on 'when' the failure happened?. Sounds weird to me and like we are mixing concerns. It's easier for my brain to pick the idea of having the supported load state in the `Index`, SIM to track that instead of trying to infer it, set the defaults we'd like and be done. But that's for another discussion as you rightfully point.

I did try that change in SIM and this sends `SIMTest` into the most amazing failure scenarios with indexes blocking when they shouldn't, jstaks on all threads blocked and all kind of funnies. I did spend some time on it without success until I realized it should be:

`return isInitialBuild ? LoadType.WRITE : LoadType.READ;`

Correct?

adelapena replied:
Hi @bereng. I understand that the original behaviour was:
```java
return isInitialBuild ? LoadType.WRITE : LoadType.ALL;
```
That's because the indexes were always writable independently of any kind of build failures, there wasn't even a notion of writability. And, once they were queryable once, they were queryable forever.

Just took a quick look at the `SIMTest` failures, I think they are happening because `TestingIndex.shouldFailCreate` is wrongly missed from `TestingIndex.clear()`. Once it's fixed to reset it, and we change a couple of assertions about writability in `initializingIndexNotQueryableButWritableAfterPartialRebuild`, all the tests seem to pass, at least locally.

I'll take a closer look tomorrow.

bereng replied:
Ah good catch! you're right that is being missed. Amazing this didn't came up until today.

In any case I'd like your feedback on the extra flush also. Feel free to shoot a half-backed push if you fancy and when I get the rest of the feedback I'll adjusts tests, dtests and that should be it hopefully.

adelapena replied:
I think that the flush should happed right after we have ensured that the indexes are marked as writable. Otherwise, we might have new untracked writes populating the new memtable while the indexes are not writable. In the same vein, the sstable selection should happen after we have marked the indexes as writable. I have almost ready a commit with this and a few minor suggestions, I'll push it tomorrow. 

We're almost there!

bereng commented:
@adelapena I have pushed the latest + you can find the dtests PR on the jira ticket. The 2i dtest passes locally also. If you like what you see I'll run the full dtest suite. So far utests look good:
- [CI j11](https://app.circleci.com/pipelines/github/bereng/cassandra/25/workflows/fa67e701-2df7-42a4-9a8e-1edee5200a6c)
- [CI j8](https://app.circleci.com/pipelines/github/bereng/cassandra/25/workflows/429f4fea-de2e-434a-b5a2-99f13b5013ae)

adelapena commented:
I have left a few suggested changes [here](https://github.com/adelapena/cassandra/commit/dc0030ea4e961a2f17ebf8cc486963eec0c1e037). Changes for the dtest are [here](https://github.com/adelapena/cassandra-dtest/commit/758508b7759dfe13af724df68ee80834662ac013).

As mentioned in my previous comment, the sequence of operations in `SIM#rebuildIndexesBlocking` is modified to:
1. Set indexes as writable to not miss incoming writes.
2. Flush the base table if some index wasn't writable before, so the potentially unindexed memtable contents are moved to the set of sstables to be indexed.
3. Acquire sstable refs and rebuild.

I have also done some modifications to the way we determine whether we should use `Index#getBuildTaskSupport` or `Index#getRecoveryTaskSupport`. `SIM#buildIndexesBlocking` is only called by either `SIM#rebuildIndexesBlocking` or `SIM#handleNotification`. I understand that if the caller is `rebuildIndexesBlocking` we are always doing a full rebuild, and probably the only reason to do that is that we are attempting to recover a damaged index. Conversely, if the caller is `handleNotification` we are doing a partial build of new data that is unable to recover previous errors. This is why I think that the `isFullRebuild` argument is enough to determine whether we are recovering or indexing new data. This makes full rebuild independent of the status of the index, so any call to `nodetool rebuild_index` will involve `getRecoveryTaskSupport`, which is intended to do any initialization work that might be needed prior to a full index build. Do you think this assumption is correct, or are we interested in treating separately full rebuilds without a previous failure?

Unit tests are also slightly modified to reflect the changes. `SecondaryIndexManagerTest` is extended with queryability/writability checks over a write-only-on-failure index.

Summarizing, I think that at this point the patch achieves the following:
* Full index rebuilds give indexes the opportunity to run specific rebuild logic, including initialization. This logic can be different from the logic that is used when regularly indexing new sstables.
* If any full rebuild or a partial build task fails, the index can start to reject queries and ignore writes until a full rebuild is successfully run. Whether it rejects queries or ignores writes is defined by the index implementation.
* Rebuilt indexes don't miss data after having been ignoring writes.
* Existing indexes preserve their original behaviour. They always accept writes and only reject reads if the initial build has failed.
* The index API is backwards compatible.

Do you think it makes any sense?

bereng commented:
Hi @adelapena. Thx for the commit. I patched it and apply it here. Feel free to push here directly next time so it's authored by you rather than me so I don't steal your thunder :-)

I think what you do here makes perfect sense from the pov you have. Basically you try to infer as much as possible and rely on reasonable assumptions. I still think, call me stubborn :-) , I'd like my approach better. But as you mention this patch brings goodness and if in the future we need to move closer to my pov it's an easy change. So I am +1 on merging this. Having catched the missing flush, the extra testing around things, etc it's all good value.

Let me run CI and I'll post results here when done.

bereng commented:
CI runs:
- [CI j8](https://app.circleci.com/pipelines/github/bereng/cassandra/34/workflows/aca7db2f-b52d-452d-8264-c0a478c11a76)
- [CI j11](https://app.circleci.com/pipelines/github/bereng/cassandra/34/workflows/82d75275-3e69-4c2f-b26e-4369563ecfd4)

I checked failures: SlowQueryLog is a known one, the SimpleStragey cast one is one I fixed in another ticket, the Cqlsh one on empty values passes locally, the other few look completely unrelated, etc. I _think_ we're good to merge.

adelapena commented:
@bereng I can't see the links to CI. I'm running [ci-cassandra](https://ci-cassandra.apache.org/job/Cassandra-devbranch/133/) too. Anyway, it seems that we are mostly ready to commit.

> I think what you do here makes perfect sense from the pov you have. Basically you try to infer as much as possible and rely on reasonable assumptions. I still think, call me stubborn :-) , I'd like my approach better. But as you mention this patch brings goodness and if in the future we need to move closer to my pov it's an easy change. So I am +1 on merging this. Having catched the missing flush, the extra testing around things, etc it's all good value.

Now that we have the general mechanism to define queryability and writability on failure, we can open a followup ticket to more aggressively change the default behaviour of Cassandra indexes, possibly with some discussion in the mail list so we don't interfere with existing use cases. If necessary, it wouldn't be difficult to pass the `LoadType` to the CQL index options, so users can decide how their indexes will behave, choosing between availability and consistency. As for failing writes on badly built indexes instead of ignoring them, we should also have some discussion because it will quite aggressively affect the base table/keyspace availability, messing with repairs, compactions, etc. It will also might make sense to have some mechanism to make coordinators aware of which nodes are not available for reads and writes so they can avoid them.

bereng commented:
Great, so you can't access them. I can't tell why as they seem to work for me and your ci-cassandra run failed on trying to pull my dtests branch. The joys of CI :-) What error do you get on my links or do you want to rerun yours?

adelapena commented:
I see a message telling `Workflows >> null >> null >> null >>`. However I can see the list of your workflows on https://circleci.com/gh/bereng/workflows/cassandra , but I can't see the detail of any of them 🤷

Currently I'm having trouble to login at ci-cassandra and re-run, I'll try later.

adelapena commented:
I have managed to see your CI run. It seems that CircleCI was punishing me for using the old UI, and at the same time it wasn't showing me the option to upgrade to the new UI 🤦 

The results look good to me to. `SlowQueryLog` is failing because dtests include CASSANDRA-15503 but this PR doesn't. Perhaps we should squash+rebase and run a final CI round, and we'll be ready to merge, WDYT?

bereng commented:
@adelapena apologies for the late reply. I was going down a rabbit hole debugging a failing test yesterday and I didn't get around to this. Rebased and squashed. We'll have to wait for the CI runs to complete now:
- [CI j8](https://app.circleci.com/pipelines/github/bereng/cassandra/35/workflows/54e22147-7fb8-4171-a0be-1f189b9f657d)
- [CI j11](https://app.circleci.com/pipelines/github/bereng/cassandra/35/workflows/4640daf9-6499-499a-9dab-4eb15fac85db)

bereng commented:
@adelapena The CI runs are complete. The failures I see I think are just flaky tests because I ran some locally and they passed.

The scare comes from the j8/j11UnitTests that fail on `SIMT` (yikes). So I ran that test locally compiled in j8 under j8 and it passes. It passes also compiled on j8 and ran under j11. I checked the logs for the test and it's full of socket and 'too many open files' exceptions. I did run `ant test` under j11 locally and all passed.

Also I managed to rerun that j8 run [here](https://circleci.com/workflow-run/e1f63b50-0393-46ac-a1c4-c529a4ad9907) and you can see now a j8 utest failed but the j11 utests all passed :shrug:  (pick your poison lol)

I think this is all CI headaches/noise. It would be great if you could run j8/j11 utests yourself on cassandra ci before merging to be on the safe side. Wdyt?

adelapena commented:
> I think this is all CI headaches/noise. It would be great if you could run j8/j11 utests yourself on cassandra ci before merging to be on the safe side. Wdyt?

@bereng Most probably it's just CI noise. I'm running cassandra-ci, now that I'm able to login into it again:
|utest|dtest|
|-----|-----|
|[108](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-test/108/)|[126](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-dtest/126/)|

Unfortunately I don't know if there is a way to specify the Java version in cassandra-ci.

One of the errors seems to occur in `SecondaryIndexManagerTest`. We might have made it flaky, so I'm running our classic private multiplexer [here](https://jenkins-cassandra.datastax.lan/view/Parameterized/job/parameterized-testall/650/) and [here](https://jenkins-cassandra.datastax.lan/view/Parameterized/job/parameterized-testall/651/) 🤞

adelapena commented:
@bereng It seems that `SIMT#initializingIndexNotQueryableButMaybeNotWritableAfterPartialRebuild` and `indexWithFailedInitializationDoesNotChangeQueryabilityNorWritabilityAfterPartialRebuild` are definitively flaky. I haven't dug into it, but it seems that somehow the SIM calls `logAndMarkIndexesFailed` with an unexpected null exception. The failure can be reproduced just running any of the problematic tests multiple times in IntelliJ, or you can see them in the multiplexer runs above.

bereng commented:
@adelapena ok that took a while to pin down. There were concurrency issues with the sets keeping track of the indices, which lead to wrong `failedIndexes` lists triggering the NPE. The fix is [here](https://github.com/apache/cassandra/commit/dd0a598c8fa29b3bfcda1ec66c9190904a84b10d#diff-3f2c8994c4ff8748c3faf7e70958520dR489). It was a legit bug :-)

I did run the test 1K times and it passed. Now waiting on CI.

Edit: Started CI again upon a GH connection error that failed them...

adelapena commented:
Nice detective work. So it seems that by introducing multiple indexes in the test we have found and fixed and unrelated bug in the SIM, that's a good thing. 

If CI looks we'll finally be ready to merge :shipit:

bereng commented:
@adelapena circle CI is on a partial outage and it fails on dtests env set up etc. If you fancy firing one in cassandra ci good, otherwise I'll try to get CI runs again tomorrow #justfyi

adelapena commented:
Cassandra CI re-running [utests](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-test/112/) and [dtests](https://ci-cassandra.apache.org/view/all/job/Cassandra-devbranch-dtest/134/).

bereng commented:
Lgtm. Wdyt?

adelapena commented:
Looks good to me too, merging.

adelapena commented:
Committed to master as [595dc61290a9fda15b6765711141039ec9609bb3](https://github.com/apache/cassandra/commit/595dc61290a9fda15b6765711141039ec9609bb3).
We can close this PR and the one for dtest.
